collectors/modality-probe-debug-collector/src/lib.rs

```diff
@@ -0,0 +1,1169 @@
+            Word::U64(n) => Word::U64(n + rhs),
+        }
+    }
+}
+
+// Offset of pointer to DynamicHistory in ModalityProbe struct, which is located in modality-probe/src/lib.rs
+fn history_ptr_offset() -> u64 {
+    offset_of!(ModalityProbe => history).get_byte_offset() as u64
+}
+
+// Address offsets of each needed field of the DynamicHistory struct, which is located in modality-probe/src/history.rs
+fn overwrite_priority_offset() -> u64 {
+    offset_of!(DynamicHistory => overwrite_priority).get_byte_offset() as u64
+}
+
+fn probe_id_offset() -> u64 {
+    offset_of!(DynamicHistory => probe_id).get_byte_offset() as u64
+}
+
+fn write_seqn_high_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => write_seqn: SeqNum => high)
+        .get_byte_offset() as u64
+}
+
+fn write_seqn_low_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => write_seqn: SeqNum => low)
+        .get_byte_offset() as u64
+}
+
+fn overwrite_seqn_high_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => overwrite_seqn: SeqNum => high)
+        .get_byte_offset() as u64
+}
+
+fn overwrite_seqn_low_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => overwrite_seqn: SeqNum => low)
+        .get_byte_offset() as u64
+}
+
+fn log_storage_addr_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => storage).get_byte_offset()
+        as u64
+}
+
+fn log_storage_cap_offset(n_word_bytes: u8) -> u64 {
+    // Assume storage addr is a u32
+    log_storage_addr_offset() + n_word_bytes as u64
+}
+
+/// Configuration for running collector
+#[derive(Debug, PartialEq)]
+pub struct Config {
+    pub session_id: SessionId,
+    pub target: Target,
+    pub interval: Duration,
+    pub output_path: PathBuf,
+    pub probe_addrs: Vec<ProbeAddr>,
+}
+
+/// Target device, either directly through probe-rs or by proxy through a gdb server
+#[derive(Debug, PartialEq)]
+pub enum Target {
+    ProbeRsTarget(String),
+    GdbAddr(SocketAddrV4),
+}
+
+/// Struct representing a probe address, either the address of the probe itself or of
+/// a pointer to the probe
+#[derive(Debug, PartialEq)]
+pub enum ProbeAddr {
+    Addr(Word),
+    PtrAddr(Word),
+}
+
+/// Error in debug collector
+#[derive(Debug, Error)]
+pub enum DebugCollectorError {
+    #[error(display = "Error reading from/writing to target")]
+    TargetError,
+    #[error(display = "Invalid probe id read from device")]
+    ProbeIdError,
+    #[error(display = "Error processing the report's log")]
+    LogProcessingError,
+    #[error(display = "Error serializing the report")]
+    ReportSerializationError,
+    #[error(display = "Error using output file")]
+    FileError,
+}
+
+/// Trait used to specify backend used to access device memory
+pub trait MemoryAccessor {
+    fn read_word(&mut self, addr: Word) -> Result<Word, DebugCollectorError>;
```

Comment:
It seems like the implementations of this trait only ever return the `DebugCollectorError::TargetError` variant of `DebugCollectorError`. I recommend changing the error-branch return type of this trait to something specific to the kinds of failures that are actually expected to occur in the trait implementations. Then, as necessary, at the call site or in a `From` implementation, map that trait-specific error return to the appropriate higher-level error type (i.e. `DebugCollectorError`) variants.